### PR TITLE
Add Intro documentation for Java API (#454)

### PR DESCRIPTION
**Pull request by jlprat**
Issue: #454
Add Java snippets for:

- Minimal Server (#minimal-routing-example)
- Spray JSON example (#second-spray-json-example)
- Streaming Random Numbers (#stream-random-numbers)

Missing Java snippets for:
- Actor Interaction (#actor-interaction)
- Low Level Server Example (#actor-interaction)
- Single Client Request (#single-request-example)

### CODE REVIEW

**Comment by jlprat** [comment]
Note:
As all examples should be complete (including imports) I put each example as a separated file so it can be included nicely.
Unfortunately, in Java I can't add imports in the middle of the class ;)

**Comment by ktoso** [comment]
OK TO TEST

**Comment by jonas** [comment]
It is possible to include multiple snippets in Paradox, either using:
```java
// Java
//#imports
import akka.actor.ActorSystem;
//#imports
...
class ExampleTest {
  //#usage
  final ActorSystem system = ActorSystem.create();
  //#usage
}

// Docs
@@snip [ExampleSpec](.../ExampleSpec.java) { #imports #usage }
```
Or using the same snippet ID:
```java
// Java
//#some-example
import akka.actor.ActorSystem;
//#some-example
...
class ExampleTest {
  //#some-example
  final ActorSystem system = ActorSystem.create();
  //#some-example
}

// Docs
@@snip [ExampleSpec](.../ExampleSpec.java) { #imports #usage }
```

For example see the [HttpClientExampleDocTest.java](https://github.com/akka/akka-http/blob/master/docs/src/test/java/docs/http/javadsl/HttpClientExampleDocTest.java) used [here](http://doc.akka.io/docs/akka-http/current/java/http/implications-of-streaming-http-entity.html#consuming-the-http-response-entity-client-).

One improvement we might want to do in Paradox is to have an option to "undent" all snippets to the same level so the example above doesn't show:

```java
import akka.actor.ActorSystem;
  final ActorSystem system = ActorSystem.create();
``

**Comment by jlprat** [comment]
That's a cool feature, I didn't know!
However, I don't know if I would be wise to merge those examples in the same file as somebody hitting "reorganize imports" might change the order of the imports causing bad examples.

**Comment by jonas** [comment]
True, I've been thinking of adding Paradox support for `@@@ snip` blocks which would generate a test file and compile/run it for this sort of thing and to make it simpler to write small testable examples.

**Comment by jlprat** [comment]
Added 3 remaining snippets.
Couldn't test paradox generation due to some strange error. Will do tomorrow

File: docs/src/main/paradox/java/http/index.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ akka-http-jackson
 
 @@@ index
 
+* [introduction](introduction.md)
```
Comment (ktoso) [inline docs/src/main/paradox/java/http/index.md]: 👍

**Comment by jlprat** [comment]
This PR is not a WIP anymore :)

File: docs/src/main/paradox/java/http/introduction.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+However, if your application is not primarily a web application because its core is not browser-interaction but some specialized maybe complex business service and you are merely trying to connect it to the world via a REST/HTTP interface a web-framework might not be what you need. In this case the application architecture should be dictated by what makes sense for the core not the interface layer. Also, you probably won’t benefit from the possibly existing browser-specific framework components like view templating, asset management, JavaScript- and CSS generation/manipulation/minification, localization support, AJAX support, etc.
+
+Akka HTTP was designed specifically as “not-a-framework”, not because we don’t like frameworks, but for use cases where a framework is not the right choice. Akka HTTP is made for building integration layers based on HTTP and as such tries to “stay on the sidelines”. Therefore you normally don’t build your application “on top of” Akka HTTP, but you build your application on top of whatever makes sense and use Akka HTTP merely for the HTTP integration needs.
```
Comment (jlprat) [inline docs/src/main/paradox/java/http/introduction.md]: Sounds good, this should be added to both docs (Java and Scala) then. Would it be better to have a separated issue for this?

Comment (ktoso) [inline docs/src/main/paradox/java/http/introduction.md]: In both I think :) Thanks! Will review the rest in a bit.

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
PLS BUILD

**Comment by jlprat** [comment]
Did I break any test?

**Comment by jonas** [comment]
Some of the tests are rather timing sensitive which means that they sometimes fail if the Jenkins servers are overloaded. When this happens a ticket with the `failure` label is created to track it. In this case there's already a ticket: #493.

File: docs/src/main/paradox/java/http/introduction.md (hunk 2 above)
Comment (jlprat) [inline docs/src/main/paradox/java/http/introduction.md]: So, should I add this text under this PR? or better have a new issue + PR for better traceability?

Comment (ktoso) [inline docs/src/main/paradox/java/http/introduction.md]: Here is fine, thanks!

Comment (jlprat) [inline docs/src/main/paradox/java/http/introduction.md]: Just pushed, let's see if you like the wording

File: docs/src/test/java/docs/http/javadsl/SprayJsonExampleTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  private Route createRoute() {
+
+    return route(
+      get(() ->
+        pathPrefix("item", () ->
+          path(longSegment(), (Long id) -> {
+            final CompletionStage<Optional<Item>> futureMaybeItem = fetchItem(id);
+            return onSuccess(() -> futureMaybeItem, maybeItem ->
+              maybeItem.map(item -> completeOK(item, Jackson.marshaller()))
```
Comment (ktoso) [inline docs/src/test/java/docs/http/javadsl/SprayJsonExampleTest.java]: This is not using spray 😉  It's using Jackson, all our docs examples for java (and support) is for jackson

File: docs/src/main/paradox/java/http/introduction.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+akka-http-core
+: A complete, mostly low-level, server- and client-side implementation of HTTP (incl. WebSockets)
+Details can be found in sections @ref[Low-Level Server-Side API](server-side/low-level-server-side-api.md#http-low-level-server-side-api) and @ref[Consuming HTTP-based Services (Client-Side)](client-side/index.md#http-client-side)
+
+akka-http-testkit
+: A test harness and set of utilities for verifying server-side service implementations
+
+akka-http-spray-json
+: Predefined glue-code for (de)serializing custom types from/to JSON with [spray-json](https://github.com/spray/spray-json)
+Details can be found here: @ref[JSON Support](common/json-support.md#akka-http-spray-json)
```
Comment (ktoso) [inline docs/src/main/paradox/java/http/introduction.md]: This should be removed - java does not care about spray :-)
See: See https://github.com/akka/akka-http/pull/521/files#r88125838

Toe example we want to point to here is jackson

**Comment by ktoso** [comment]
One small thing to fix about jackson/spray-json and then LGTM :-)
Thanks a lot for converting these!

**Comment by jlprat** [comment]
I will change it later today. I didn't even think about renaming the examples to Jackson, even if it was quite obvious...

**Comment by ktoso** [comment]
Thanks, and just in time for tomorrow's release :)

**Pull request merged**